src/GitHub.App/Services/TeamExplorerContext.cs

```diff
@@ -0,0 +1,221 @@
+    /// on every IGitExt property change event). <see cref="ActiveRepository"/> contains the active repository or null
+    /// if a solution is opened that isn't in a repository. No events are fired when the same solution is unloaded then
+    /// reloaded (e.g. when a .sln file is touched).
+    /// </remarks>
+    [Export(typeof(ITeamExplorerContext))]
+    [PartCreationPolicy(CreationPolicy.Shared)]
+    public class TeamExplorerContext : ITeamExplorerContext
+    {
+        const string GitExtTypeName = "Microsoft.VisualStudio.TeamFoundation.Git.Extensibility.IGitExt, Microsoft.TeamFoundation.Git.Provider";
+
+        readonly ILogger log;
+        readonly DTE dte;
+
+        string solutionPath;
+        string repositoryPath;
+        string branchName;
+        string headSha;
+        string trackedSha;
+
+        readonly IService service = new Service();
+
+        [ImportingConstructor]
+        public TeamExplorerContext(IGitHubServiceProvider serviceProvider)
+            : this(LogManager.ForContext<TeamExplorerContext>(), new Service(), GitExtTypeName, serviceProvider)
+        {
+        }
+
+        public TeamExplorerContext(ILogger log, IService service, string gitExtTypeName, IGitHubServiceProvider serviceProvider)
+        {
+            this.log = log;
+            this.service = service;
+
+            // Visual Studio 2015 and 2017 use different versions of the Microsoft.TeamFoundation.Git.Provider assembly.
+            // There are no binding redirections between them, but the package that includes them defines a ProvideBindingPath
+            // attrubute. This means the required IGitExt type can be found using an unqualified assembly name (GitExtTypeName).
+            var gitExtType = Type.GetType(gitExtTypeName, false);
+            if (gitExtType == null)
+            {
+                log.Error("Couldn't find type {GitExtTypeName}", gitExtTypeName);
+            }
+
+            var gitExt = serviceProvider.GetService(gitExtType);
+            if (gitExt == null)
+            {
+                log.Error("Couldn't find service for type {GitExtType}", gitExtType);
+                return;
+            }
+
+            dte = serviceProvider.TryGetService<DTE>();
+            if (dte == null)
+            {
+                log.Error("Couldn't find service for type {DteType}", typeof(DTE));
+            }
+
+            Refresh(gitExt);
+
+            var notifyPropertyChanged = gitExt as INotifyPropertyChanged;
+            if (notifyPropertyChanged == null)
+            {
+                log.Error("The service {ServiceObject} doesn't implement {Interface}", gitExt, typeof(INotifyPropertyChanged));
+                return;
+            }
+
+            notifyPropertyChanged.PropertyChanged += (s, e) => Refresh(gitExt);
+        }
+
+        /// <summary>
+        /// Used for unit testing.
+        /// </summary>
+        public interface IService
```

Comment:
What is this service interface for?

Reply:
To make it unit testable. I didn't want to create an external, public service that wouldn't be used by anything else. It isn't pretty, but I couldn't think of a less hacky way.